### ISSUE TITLE
Error loading state after Terraform upgrade

### ISSUE THREAD

**Issue by westredd** [hashicorp/terraform#15380]
I am trying to upgrade from terraform 0.9.3 to 0.9.8
After changing the terraform.exe file to 0.9.8 a terraform plan fails with a state file error
I also tried upgrading from 0.9.3 to 0.9.6 but the same issue appears.

My required version field is as follows.

required_version = ">= 0.9.3"

If i revert the terraform.exe back to 0.9.3 the problem goes away.

### Terraform Version
0.9.8

### Expected Behavior
Expect a terraform plan to run

### Actual Behavior
Output below

Error loading state: state data in S3 does not have the expected content.

This may be caused by unusually long delays in S3 processing a previous state
update.  Please wait for a minute or two and try again. If this problem
persists, and neither S3 nor DynamoDB are experiencing an outage, you may need
to manually verify the remote state and update the Digest value stored in the
DynamoDB table to the following value: 

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Change the terraform version from 0.9.3 to 0.9.8
2. Run a terraform plan

### Important Factoids
No

### References
None

**Comment by jbardin** [comment]
Hi @westredd,

Sorry this is causing trouble for you, but I'm not sure how you arrived at this error.

Terraform 0.9.3 didn't yet track the checksums of S3 objects, so going directly from 0.9.3 -> 0.9.8 shouldn't produce the error, since there is no stored checksum to fail verification. The state file seems to have been modified _after_ running a refresh or apply with 0.9.8, in which case terraform is reporting the error correctly.

Are there any intermediate steps you took that may have re-written the state file outside of Terraform 0.9.8? Is it possible you ran Terraform 0.9.8 without the dynamodb entry in the config at one point?

**Comment by kennethjiang** [comment]
We ran into the same error with 0.9.8 and haven't been able to figure it out for a few days now!

For us, however, the error didn't happen immediately after we upgraded to 0.9.8. Instead it happened when one `terraform apply` seem to screw up remote state big time.

When we switched back to 0.9.6, we got a different error that seems to be slightly more descriptive:

```
2017/06/23 10:11:53 [INFO] backend/local: starting Plan operation
2017/06/23 10:11:54 [DEBUG] plugin: waiting for all plugin processes to complete...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x46f62aa]

goroutine 7591 [running]:
github.com/hashicorp/terraform/backend/remote-state/s3.(*RemoteClient).Get(0xc420239880, 0x11, 0x11, 0x18)
	/private/tmp/terraform-20170525-64528-ddymdk/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/remote-state/s3/client.go:64 +0x11a
github.com/hashicorp/terraform/state/remote.(*State).RefreshState(0xc420f53d60, 0x100ed72, 0xc420a29f00)
	/private/tmp/terraform-20170525-64528-ddymdk/terraform-0.9.6/src/github.com/hashicorp/terraform/state/remote/state.go:33 +0x37
github.com/hashicorp/terraform/state.(*BackupState).RefreshState(0xc420f74900, 0x5e21a2f, 0x7)
	/private/tmp/terraform-20170525-64528-ddymdk/terraform-0.9.6/src/github.com/hashicorp/terraform/state/backup.go:21 +0x33
github.com/hashicorp/terraform/backend/local.(*Local).context(0xc420c9dd60, 0xc4209c36b0, 0x2, 0x2, 0x3, 0xc420a29f00, 0x2)
	/private/tmp/terraform-20170525-64528-ddymdk/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend_local.go:31 +0x97
github.com/hashicorp/terraform/backend/local.(*Local).opPlan(0xc420c9dd60, 0x88caa60, 0xc420012580, 0xc4209c36b0, 0xc420f5ac30)
	/private/tmp/terraform-20170525-64528-ddymdk/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend_plan.go:57 +0x22f
github.com/hashicorp/terraform/backend/local.(*Local).(github.com/hashicorp/terraform/backend/local.opPlan)-fm(0x88caa60, 0xc420012580, 0xc4209c36b0, 0xc420f5ac30)
	/private/tmp/terraform-20170525-64528-ddymdk/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend.go:243 +0x52
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc420c9dd60, 0xc420792dc0, 0xc420792db0, 0x88caa60, 0xc420012580, 0xc4209c36b0, 0xc420f5ac30)
	/private/tmp/terraform-20170525-64528-ddymdk/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend.go:265 +0x9a
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/private/tmp/terraform-20170525-64528-ddymdk/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend.go:266 +0x192
```

Suggestions on how we can work around it will be highly appreciated since our whole engineering team is blocked on this.

**Comment by westredd** [comment]
> Are there any intermediate steps you took that may have re-written the state file outside of Terraform 0.9.8? 

A couple of weeks ago i also tried to upgrade.  I ran a TF plan and it was showing as adding double the resources that it actually needed to. So for example, if i was adding 1 resource, it would show 1 resourrce but in the summary it would say "2 to add".  And if there were 5 resources to add it would say "10 to add" so i went back to 0.9.3.

So today i went back to trying the upgrade again and got the error written in my post.  Any ideas how i can resolve?

**Comment by jbardin** [comment]
@westredd, thanks, that's helpful. While a plan doesn't modify the state by default, it may have somehow updated the state file metadata in the backend. When you moved back to 0.9.3 the state file changed and took it out of sync with the DynamoDB entry. If that's the case, I'll see if there's a way we can prevent the checksum from being written if the state isn't modified. 

@kennethjiang, that crash is a bug fixed since 0.9.6, but the workaround is the same.

The workaround here for both these cases is to remove the out-of-sync entry in DynamoDB. There will be a LockID entry in the table containing state and expected checksum which you can delete.

**Comment by westredd** [comment]
Thanks for that.

Have you got any idea why the resources  were summarised as 2 times what was due to be added?  Has anyone else mentioned that?  I have 6 o 7 state files that use moduled code and all were showing 2 times the resource to be added in a TF plan. Fine on 0.9.3 but as soon as move to a higher version they show as doubled.

**Comment by kennethjiang** [comment]
Thank you so much @jbardin . Removing lock record in DynamoDB solved the problem for us!

**Comment by westredd** [comment]
I think this "Error loading state: state data in S3 does not have the expected content." error is only affecting my sandbox environment state file and thinking about it i may have done a TF apply in 0.9.8 a few weeks ago against that state file. it was test and UAT, Dev that i only did a TF plan and have just realised that these envs i can still run a plan on 0.9.8. Quite a relief

I can rebuild the sandbox TF state easily enough, but any ideas why the resources show as doubled? I dont want to move to 0.9.8 if its doing that.  Also does it on 0.9.6.  haven't tried 0.9.4 or 0.9.5. Any ideas?

**Comment by apparentlymart** [comment]
@westredd the double-counting of certain resources in plan is indeed a known issue, which is being discussed in #14905.

**Comment by iceycake** [comment]
Saw the same error with 0.9.11 but only with one of our teams.  Try to find the entry in the dynamodb table but nothing.

**Comment by dthapa** [comment]
not sure if this is going to help anyone but I was getting a similar issue working through terraform up and running book, and it was during init
v0.9.11

apparently I got the key wrong - 
changed
key = "stage/services/webserver-cluster/s3/terraform.tfstate"
to
key = "stage/services/webserver/s3/terraform.tfstate"
and things worked.

**Comment by zerolaser** [comment]
@jbardin  any update on when this bug going to be fixed or any workaround ?

**Comment by westredd** [comment]
@zerolaster If you are talking about the double counting, I read on another thread that its due to be fixed on version 0.10 which I believe is the next release.

**Comment by jbardin** [comment]
@zerolaser,

Sorry, this probably should have been closed already, but it wasn't connected to a specific PR. 
I think all of the issues mentioned here have have been taken care of by various patches or workarounds. If you're still having this issue with a current release, please add the steps to reproduce and I can reevaluate the issue and reopen as necessary.
